rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+
+As powerful as the new module becomes, it is important to move basic and common functionality into the core of the router because building and maintaining custom modules should be a last resort. The router should provide a rich set of features out of the box that cover the most common use cases. Custom modules should be reserved for advanced or highly specific use cases that cannot be achieved with the built-in features of the router. Integration with third-party services, custom authentication, and advanced telemetry are examples of use cases that are well-suited for custom modules.
+
+## Proposal
+
+A developer can implement a custom module by creating a struct that implements one or more of the following interfaces:
+
+- `GatewayHooks`: Provides hooks for the gateway lifecycle, including request and response handling.
+- `SubgraphHooks`: Provides hooks for subgraph requests and responses.
+- `ApplicationHooks`: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+- `AuthenticationHooks`: Provides hooks for authentication and authorization logic.
+- `TelemetryHooks`: Provides hooks for OpenTelemetry spans and metrics.
+- `OperationHooks`: Provides hooks for parsed, normalized, and planned GraphQL operations.
+- `Module`: Provides hooks for the module lifecycle, including provisioning and shutdown.
+
+```go
+type MyModule struct{}
+
+type Subgraph struct {
+	// The name of the subgraph
+	Name string
+	// The ID of the subgraph
+	ID string
+	// The URL of the subgraph
+	URL *url.URL
+	// The schema of the subgraph
+	Schema *graphql.Schema
+}
+
+type Graph struct {
+	// The name of the graph
+	Name string
+	// The ID of the graph
+	ID string
+	// Version of the graph
+	Version string
+	// Subgraphs of the graph
+	Subgraphs []*core.Subgraph
+	// GraphQL schema of the graph
+	Schema *graphql.Schema
+}
+
+type GatewayRequest struct {
+	// The original GraphQL request with all the information like query, variables, operation name, extensions etc.
+	Request *core.GraphQLRequest
+	// The parsed, normaliazed and planned operation with all the information like name, variables, type, document representation,
+	// client name version, uploaded files, plan, normalization, persisted operation etc.
+	Operation *core.Operation
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The active graph
+	Graph *core.Graph
+	// The original HTTP request
+	Orignal *http.Request
+	// Logger for the gateway request
+	Logger *zap.Logger
+}
+
+type GatewayResponse struct {
+	// The original RouterRequest
+	Request *core.GatewayRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response with all the information like data, errors, extensions etc.
+	// This is the response that will be sent to the client and can be manipulated or replaced
+	Response *core.GraphQLResponse
+	// The original HTTP response
+	Orignal *http.Response
+}
+
+type GatewayHooks interface {
```

Comment:
I would take a slightly different approach to defining interfaces.
Instead of having one big iFace with many funcs, create multiple smaller ones and combine them.

Example:

```go
type GatewayHooks interface {
  GatewayOnRequestHook
  GatewayOnResponseHook
  GatewayOnErrorHook
}

type GatewayOnRequestHook interface {
  OnGatewayRequest(ctx *core.GatewayRequest, err error) error
}

type GatewayOnResponseHook interface {
  OnGatewayResponse(ctx *core.GatewayResponse, err error) error
}

type GatewayOnErrorHook interface {
  OnGatewayError(err error)
}
```

The reasoning is that a user can implement such hooks in a more granular way.
They might need `OnGatewayResponse` but don't want the other functions.
So an implementer only needs to implement what they really need,
and the hook system only needs to call what was implemented.

Reply:
Great idea. Except the Module interface needs to be implemented fully.